### PR TITLE
add game language to tech support output

### PR DESCRIPTION
**Pull request by wexxlee**
This adds the current game client language to the tech support clipboard output.  Tested & working with the current global client, but I can't validate that that the pre-7.0 signatures work with current `cn`/`ko` clients (nor can I determine what their language byte is), so would appreciate if anyone on those clients is able to test and provide that info.  But this will at least return values for unknown languages in the meantime, and it can always be supplemented later.

I haven't mucked around much with this part of the code before, so apologies in advance for any dumb stuff.

### CODE REVIEW

**Comment by valarnin** [comment]
I'll try to find some time to do a more in-depth review of this PR tomorrow, but as a quick initial point of feedback, `FFXIVClientStructs` structs should go in their own file (via `partial class`).

Reference the `JobGauge70.cs` and `JobGauge70ClientStructs.cs` files for an example.

This makes it easier to determine what has been pulled directly from FFXIVClientStructs, and to separate CS layouts from OP layouts/logic.

**Comment by wexxlee** [comment]
> I'll try to find some time to do a more in-depth review of this PR tomorrow, but as a quick initial point of feedback, `FFXIVClientStructs` structs should go in their own file (via `partial class`).
> 
> Reference the `JobGauge70.cs` and `JobGauge70ClientStructs.cs` files for an example.
> 
> This makes it easier to determine what has been pulled directly from FFXIVClientStructs, and to separate CS layouts from OP layouts/logic.

Sure, makes sense.  Do you have any concerns about the approach I'm taking with the struct, e.g., just defining the one offset we care about)?  Or would you prefer the entire CS struct for `Framework` (which is fairly large and has a lot of stuff we don't & won't ever need) be included as well?

**Comment by valarnin** [comment]
> Do you have any concerns about the approach I'm taking with the struct, e.g., just defining the one offset we care about? Or would you prefer the entire CS struct for `Framework` (which is fairly large and has a lot of stuff we don't & won't ever need) be included as well?

My general rule of thumb is to include the entire struct, but exclude any pointers. If we're fetching the entire struct from memory anyways (e.g. `Size = 0x35C8`, meaning a 0x35C8 byte memory read happens no matter what), we might as well include all of the simple values as part of the struct def, since it costs us nothing from a performance standpoint.

**Comment by wexxlee** [comment]
> My general rule of thumb is to include the entire struct, but exclude any pointers. If we're fetching the entire struct from memory anyways (e.g. `Size = 0x35C8`, meaning a 0x35C8 byte memory read happens no matter what), we might as well include all of the simple values as part of the struct def, since it costs us nothing from a performance standpoint.

OK, I think [c6a0975](https://github.com/OverlayPlugin/OverlayPlugin/pull/399/commits/c6a0975380c4c06978dee1af3f2bfa01eea82071) should do this correctly.  There's a lot of CS-specific types that are used in the struct, so rather than import a bunch of types we're never going to use and make this more bloated, I opted to just comment out those offsets in the struct.  Let me know if you have something different in mind though.

File: OverlayPlugin.Core/MemoryProcessors/ClientFramework/ClientFramework.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        Japanese = 0,
+        English = 1,
+        German = 2,
+        French = 3,
+        // The global client only supports the above options, but presumably
+        // other clients have different byte values for their languages.
+        // We don't know what these are yet, so make all other possible
+        // values 'Unknown', and we can map if/once they are known.
+        Unknown4 = 4,
```
Comment (MnFeN) [inline OverlayPlugin.Core/MemoryProcessors/ClientFramework/ClientFramework.cs:17]: `Chinese = 4,`

Just checked the 6.55 signature with CN client

File: OverlayPlugin.Core/MemoryProcessors/ClientFramework/ClientFrameworkMemory655ClientStructs.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
+﻿using System.Runtime.InteropServices;
+
+namespace RainbowMage.OverlayPlugin.MemoryProcessors.ClientFramework
+{
+    partial class ClientFrameworkMemory655 : ClientFrameworkMemory, IClientFrameworkMemory655
+    {
+        // Due to lack of multi-version support in FFXIVClientStructs, we need to duplicate these structures here per-version
+        // We use FFXIVClientStructs versions of the structs because they have more required details than FFXIV_ACT_Plugin's struct definitions
```
Comment (valarnin) [inline OverlayPlugin.Core/MemoryProcessors/ClientFramework/ClientFrameworkMemory655ClientStructs.cs:8]: This line comment is irrelevant, FFXIV_ACT_Plugin doesn't have a struct for this.

File: OverlayPlugin.Core/MemoryProcessors/ClientFramework/ClientFrameworkMemory70ClientStructs.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
+﻿using System.Runtime.InteropServices;
+
+namespace RainbowMage.OverlayPlugin.MemoryProcessors.ClientFramework
+{
+    partial class ClientFrameworkMemory70 : ClientFrameworkMemory, IClientFrameworkMemory70
+    {
+        // Due to lack of multi-version support in FFXIVClientStructs, we need to duplicate these structures here per-version
+        // We use FFXIVClientStructs versions of the structs because they have more required details than FFXIV_ACT_Plugin's struct definitions
```
Comment (valarnin) [inline OverlayPlugin.Core/MemoryProcessors/ClientFramework/ClientFrameworkMemory70ClientStructs.cs:8]: Same thing for this comment

**Review by valarnin — APPROVED** [review]
`Region` is also here, this might be worth including as well?

Maybe someday in the future we can get the windowed vs borderless vs fullscreen setting from `SystemConfig` instead of the way we're doing it currently, but that's beyond the scope of this PR.

File: OverlayPlugin.Core/MemoryProcessors/ClientFramework/ClientFramework.cs (hunk 1 above)
Comment (wexxlee) [reply OverlayPlugin.Core/MemoryProcessors/ClientFramework/ClientFramework.cs:17]: > `Chinese = 4,`
> 
> Just checked the 6.55 signature with CN client

Double-great!  Thanks for the info (and for confirming the 6.55 signatures work!)

**Comment by wexxlee** [comment]
> `Region` is also here, this might be worth including as well?

GMTA - I had included it in an early draft.  But for my client, regardless of language, the byte always =2.  I wasn't sure what that corresponded to, or what the other potential values could be, so wasn't sure how to go about populating an enum for it.  Do you know if there's a mapping somewhere?

**Comment by valarnin** [comment]
> GMTA - I had included it in an early draft. But for my client, regardless of language, the byte always =2. I wasn't sure what that corresponded to, or what the other potential values could be, so wasn't sure how to go about populating an enum for it. Do you know if there's a mapping somewhere?

Ah, I was hoping it corresponded to Global vs CN vs KR game client, but apparently it's the physical data center region instead. See:

https://beta.xivapi.com/api/1/sheet/WorldDCGroupType?fields=Region,Name

So not really as useful as I was hoping for.

**Comment by MnFeN** [comment]
7.0 CN client has a different signature:
image

**Comment by wexxlee** [comment]
> 7.0 CN client has a different signature: image

This looks like the same signature that we're using for the current int'l client in [ClientFrameworkMemory70.cs](https://github.com/OverlayPlugin/OverlayPlugin/blob/main/OverlayPlugin.Core/MemoryProcessors/ClientFramework/ClientFrameworkMemory70.cs).  OP should now be using the 7.0 signatures & structs for the `cn` client:
https://github.com/OverlayPlugin/OverlayPlugin/blob/5c655723eb9b909a2bedfb69a0d08b959cf5e904/OverlayPlugin.Core/MemoryProcessors/FFXIVMemory.cs#L27

Were you just noting the change, or is it not working as expected?

**Comment by valarnin** [comment]
CN (and KR) opcodes were merged and the versions were bumped in `FFXIVMemory.cs`, but there hasn't been a release so the plugin is defaulting to the older signatures still. If they're found, regardless of if they're valid or not, they'll be used.

tl;dr: We probably need a version bump.

**Comment by MnFeN** [comment]
> > 7.0 CN client has a different signature: image
> 
> This looks like the same signature that we're using for the current int'l client in [ClientFrameworkMemory70.cs](https://github.com/OverlayPlugin/OverlayPlugin/blob/main/OverlayPlugin.Core/MemoryProcessors/ClientFramework/ClientFrameworkMemory70.cs).  OP should now be using the 7.0 signatures & structs for the `cn` client:
> https://github.com/OverlayPlugin/OverlayPlugin/blob/5c655723eb9b909a2bedfb69a0d08b959cf5e904/OverlayPlugin.Core/MemoryProcessors/FFXIVMemory.cs#L27
> 
> Were you just noting the change, or is it not working as expected?

Please check the signature in the first part of my image. The second part is the sig for global client (which is the same as that one in ...Memory70.cs), but this sig is not found in 7.0 Chinese client.

I just noticed this difference and I have not tried to test this language feature in Chinese client yet, but several plugins using this sig in 7.0 global client could not work in 7.0 cn.
